aten/src/ATen/native/Integration.cpp

```diff
@@ -38,16 +39,17 @@ Tensor zeros_like_except(const Tensor& y, int64_t dim) {
 
 }
 
-Tensor trapz(const Tensor& y, const Tensor& x, int64_t dim) {
+Tensor trapezoid(const Tensor& y, const Tensor& x, int64_t dim) {
     dim = maybe_wrap_dim(dim, y);
     // asking for the integral with zero samples is a bit nonsensical,
     // but we'll return "0" to match numpy behavior.
     if (y.size(dim) == 0) {
         return zeros_like_except(y, dim);
     }
+    TORCH_CHECK(y.scalar_type() != kBool && x.scalar_type() != kBool, "trapezoid: received a bool input for `x` or `y`, but bool is not supported")
     Tensor x_viewed;
     if (x.dim() == 1) {
```

Comment:
I don't think this is what NumPy does, see

https://github.com/numpy/numpy/blob/3d83143c2c2f877ee921c4c7c6746d836c459839/numpy/lib/function_base.py#L4228

NumPy is very aggressive about reshaping `x`. For example:

```
y = np.array(((1, 2, 3), (4, 5, 6)))
x = np.array((1, 1))

np.trapz(y, x)
: array([0., 0.])

np.trapz(y, x, axis=0)
: array([0., 0., 0.])
```

This is because NumPy reshapes the results of `diff`ing `x` when `x` is one dimensional. 

I think it's OK to be more restrictive than NumPy as long as we're principled about it, but we should be clear about the deviation

Reply:
Line 57 below let's `x_viewed` be `x` when `x` has zero dimensions or two or more dimensions. Is there a check that should be done on the size of `x` in these cases? In general, is the rule that if `x` is 1d it's interpreted as the spacing along the specified axis, and then that tensor is reshaped/broadcast to the shape of `y`, and if `x` is not 1d then `x` is simply broadcast/reshaped to the shape of `y`?

Reply:
I will address these points separately:

1. Reshaping Differences (your first comment)
I agree that there is a difference in behavior here, but I think it only matters for the case where the specific `dim` of `x` has length 2 and the `dim` of `y` > 2. For example:

```
y = torch.tensor([0, 1, 2, 3])
x = torch.tensor([0, 1])
print(np.trapz(y.numpy(), x.numpy())) # Returns 4.5
print(torch.trapz(y, x)) # Fails
```
More importantly, this seems to be a conceptual different. The current implementation in torch is broadcasting `x` and numpy is broadcasting `dx` as you described. I would guess it was unintended?

**Let me know if I should make a code change to fix this. This may lead to backward compatibility issues with `trapz`, or `trapezoid` will start behaving differently from `trapz` for this case.**

2. "In general, is the rule that if `x` is 1d it's interpreted as the spacing along the specified axis, and then that tensor is reshaped/broadcast to size of `y` along the dim in question"

Yes, this is true and I think this behavior is fine (except for the case discussed in bullet 1 above).

3.  "and if `x` is not 1d then only the dimension dim of `x` is considered and it's also reshaped/broadcast to the size of `y` along the `dim` in question?"

No, if `x` is not 1d, all dimension of `x` is considered, not just the specified `dim`. For example, this fails for both numpy and torch:

```
y = torch.tensor([[0, 1, 2], [3, 4, 5], [6, 7, 8]])
x = torch.tensor([[0, 1, 2], [3, 4, 5]])
print(np.trapz(y.numpy(), x.numpy())) # Fails
print(torch.trapz(y, x)) # Fails
```
This fails in torch because this happens in the `trapezoid` function, the call to `.slice()` only decrease the dimension of `x` along `dim` by 1, but the other dimension values are untouched.
```
// Input x with dimension (2x3)
Tensor x_left = x_viewed.slice(dim, 0, -1); // becomes (2x2)
Tensor x_right = x_viewed.slice(dim, 1); // becomes (2x2)
Tensor dx = x_right - x_left; // (2x2)
```
Later on, when `dx` is passed into `do_trapezoid`:
```
// Input y with dimension (3x3)
Tensor left = y.slice(dim, 0, -1); // becomes (3x2)
Tensor right = y.slice(dim, 1); // becomes (3x2)
return ((left + right) * dx).sum(dim) / 2.; // throws an error because dimensions of dx (2x2) and left (3x2) don't match and a broadcast isn't possible
```
In summary, all the broadcasting takes place in `(left + right) * dx` of `do_trapezoid`. An error returns if the dimensions do not match during `(left + right) * dx`.

`x_view` within `trapezoid` merely turns a 1D `x` with dimension `(n,)` into dimension `(1,1,...,n,...,1,1)` based on `dim` and `y.dim()` so that `x` can be broadcasted at a later time.

Reply:
Sorry for the long block of text, in the meantime, I am leaving the code unchanged.

Let me know if you have any suggestion.

Reply:
Thanks for the analysis. Adding a comment explaining the behavior and citing the (slight) discrepancy with NumPy would be nice, too